compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td

```diff
@@ -678,6 +678,117 @@ def IREELinalgExt_AttentionOp : IREELinalgExt_Op<"attention",
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// Im2col
+//===----------------------------------------------------------------------===//
+
+def IREELinalgExt_Im2colOp : IREELinalgExt_Op<"im2col",
+    [DeclareOpInterfaceMethods<ReifyRankedShapedTypeOpInterface>]> {
+  let summary = "Im2col operation for convolutions";
+  let description = [{
+    Im2col op for convolutions. The operation performs a transformation on the
+    input to convert it from a convolution input to an equivalent gemm input.
+    The op is defined by its input, output, some conv metadata, and some
+    indexing metadata. The `strides`, `dilations`, and `kernel_size` are taken
+    from the convolution from which this op is generated, and they define how
+    the input operand is indexed when the operation is decomposed. The shape of
+    the output should be `tensor<BxMxK>`, and the `m_pos`, `k_pos`, and
+    `batch_pos` indicate which input dimensions map to which output dimensions.
+    The `k_offset` is an offset within the output K dimension from which the
+    iteration space of the operation begins. This is used for tiling, since the
+    tiled implementation must leave the output K dimension untiled. Similarly,
+    `m_offset` is the offset within the output M dimension from which the
```

Comment:
Hmmm, I am confused by `m_offset` and `k_offset`....

Reply:
I am not sure how to tile this op generally without keeping the full view of the `H`, `W`, and `C` dimensions. The way this is tiled is that the output tensor is tiled, but the input tensor is only tiled along  the `N` dimension. This means the op needs to keep track of the offsets of the output slice to determine the mapping of the output slice to the eventual decomposed input slice, which is done with `m_offset` and `k_offset`. Basically, the `m_offset` and `k_offset` are saying "what is the offset of the tiled result tensor into the original untiled result tensor?". For example:

```
%7 = iree_linalg_ext.im2col
    strides = [1, 1] dilations = [1, 1] kernel_size = [3, 3]
    m_offset = [%arg3] k_offset = [%arg5]
    batch_pos = [0] m_pos = [1, 2] k_pos = [3]
    ins(%extracted_slice : tensor<1x34x34x640xf32>)
    outs(%extracted_slice_0 : tensor<1x8x4xf32>) -> tensor<1x8x4xf32>
```

`%arg3` and `%arg5` are the loop induction variables for `M` and `K` dimensions of the output respectively. If this op were decomposed now, then the op would loop over the `B=1` and `M=8` dimensions of the output, and create an extract and insert from the input to the output. The offsets into the input tensor would be computed using these `m_offset` and `k_offset` as something like `[B, M]->[B, (M+m_off)/32+k_off/(640*3), (M+m_off)%32+(k_off%(640*3))/640, k_off%640]`.

The alternative would be to have the offsets belong to the input instead (like `h_offset`, `w_offset`, and `c_offset`). I decided to go with an offset on the output because it is one less input for the op to carry, and I think it makes the tiling implementation simpler, leaving the complex indexing computation to the decomposition.

Reply:
I think what would be useful is to essentially write the "loop version" of the op, i.e. if we implemented "LowerToLoops" here what would it do. Just write that in the description. Then everything should be clearer. (Sorry to be pedantic here, but this is involved and more documentation would help)